src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -134,6 +136,23 @@ def expectation(self, state, normalize=False):
                 "".format(type(state)),
             )
 
+    def expectation_from_samples(self, freq, qubit_map=None):
+        Obs = self.matrix
+        if np.count_nonzero(Obs - np.diag(np.diagonal(Obs))) != 0:
+            raise_error(NotImplementedError, "Observable is not diagonal.")
+        keys = list(freq.keys())
+        if qubit_map is None:
+            qubit_map = list(range(int(np.log2(len(Obs)))))
+        counts = np.array(list(freq.values())) / sum(freq.values())
+        O = 0
+        kl = len(qubit_map)
+        for j, k in enumerate(keys):
+            index = 0
+            for i in qubit_map:
+                index += int(k[qubit_map.index(i)]) * 2 ** (kl - 1 - i)
```

Comment:
Here you can avoid the conversion from binary to decimal if `freq` is given in decimal. If these frequencies are result of some circuit execution, the user can do `result.frequencies(binary=False)` to get this without much effort.

Even if you get the frequencies in binary, I believe you can still avoid the for loops if the operation is vectorized. For example something like
```py
ids = tuple(map(lambda x: int(x, 2), freq.keys())) # covert indices to decimal
O = np.diag(Obs)[ids] * counts
```
I would also find a better name for the variable `O` because it does not look very nice in the code. I generally avoid single letter names and capitals in local variables. `O` is even more confusing because it's very close to `0`.

Reply:
Thanks for the suggestion. The user can include a `qubit_map` that indicates which qubit each position in the counts refers to. For example, let's consider a 6 qubit circuit where we only measure qubits 1, 3 and 5 because we want to get an expected value of an observable that only acts on those qubits. We will then get counts for the states 000, 001, 010 ...111. When converting these numbers to decimal we have to take into account the `qubit_map` because the first position indicates qubit 1, the second position indicates qubit 3 and the third position indicates qubit 5. With this in mind, I do not see a clear way to do it in the way you propose. If you have any suggestion, let me know and I will implement it.

I agree that the variable names are not appropriate so I have made some changes. Let me know what you think.

Reply:
Thank you for the explanation, you are right. I think some part of this calculation can still be vectorized, even with the `qubit_map`, but I am not sure if you gain anything.

If you expect `freq` to be a result of circuit execution, in principle you could instead take the the [`CircuitResult` object](https://github.com/qiboteam/qibo/blob/c99e296693aa3ee5b54de6833bbdfe97f86c193d/src/qibo/states.py#L7), which is what circuit execution returns and contains both the frequencies and the qubit_map. This could be simpler in some cases as the user would not need to specify the qubit map twice, it would be read directly from the measurement gate in the circuit. However, what you are doing here is more flexible because it also works when frequencies do not come from a circuit.

A way to have both approaches is to define an additional method `CircuitResult.expectation_from_samples(hamiltonian)` which calls the method you defined here. So one can do
```py
result = c(nshots=100)
result.expectation_from_samples(hamiltonian)
```
I am not sure if this is useful, let me know what you think.

> I agree that the variable names are not appropriate so I have made some changes. Let me know what you think

Thanks, looks better now.